onadata/libs/tests/utils/test_logger_tools.py

```diff
@@ -236,6 +237,87 @@ def test_attachment_tracking_for_nested_repeats(self):
         self.assertEquals(instance2.json[MEDIA_ALL_RECEIVED],
                           instance2.media_all_received)
 
+    def test_replaced_attachments_not_tracked(self):
+        """
+        Test that when a submission with an attachments is made,
+        and later edited, whereby the attachment is replaced,
+        the replaced attachment is no longer tracked for that
+        submission
+        """
+        md = """
+        | survey |              |        |        |
+        |        | type         | name   | label  |
+        |        | begin repeat | images | Photos |
+        |        | image        | image1 | Photo  |
+        |        | end repeat   |        |        |
+        """
+        self._create_user_and_login()
+        self.xform = self._publish_markdown(md, self.user)
+
+        xml_string = """
+        <data id="{}">
+            <meta>
+                <instanceID>uuid:UJ5jz4EszdgH8uhy8nss1AsKaqBPO5VN7</instanceID>
+            </meta>
+            <images>
+                <image1>1300221157303.jpg</image1>
+            </images>
+        </data>
+        """.format(self.xform.id_string)
+        file_path = "{}/apps/logger/tests/Health_2011_03_13."\
+                    "xml_2011-03-15_20-30-28/1300221157303"\
+                    ".jpg".format(settings.PROJECT_ROOT)
+        media_file = django_file(
+            path=file_path, field_name="image1", content_type="image/jpeg")
+        instance = create_instance(
+            self.user.username,
+            BytesIO(xml_string.strip().encode('utf-8')),
+            media_files=[media_file])
+        self.assertTrue(instance.json[MEDIA_ALL_RECEIVED])
+        self.assertEquals(instance.json[TOTAL_MEDIA], 1)
+        self.assertEquals(instance.json[MEDIA_COUNT], 1)
+        self.assertEquals(instance.json[TOTAL_MEDIA], instance.total_media)
+        self.assertEquals(instance.json[MEDIA_COUNT], instance.media_count)
+        self.assertEquals(instance.json[MEDIA_ALL_RECEIVED],
+                          instance.media_all_received)
+        patch_value = 'onadata.apps.logger.models.Instance.get_expected_media'
```

Comment:
Why do we have to patch `get_expected_media` ?

Reply:
In an attempt to mock replacing an attachment.

When editing an instance, we receive an updated xml representaion of the instance that we then use to determine attachments for the same submission.

In this test case, since the instance already recognizes the first attachment, the updated xml would be creating a new attachment, instead of replacing the first attachment.
Patching the return value of this function allows us to use the updated xml to determine the current attachments for a specific instance.

Reply:
Okay, seems from the test we're passing an updated xml. Not quite sure what's different between how you're testing it and how a user would normally interact with the API that warrants the patch on the `get_expected_media` ?

_I might be missing something here_